From b56ddb02bb1e242549199a10c204cb2699860c89 Mon Sep 17 00:00:00 2001
From: Kiril Mikos <kiril.mik.os@gmail.com>
Date: Sun, 30 Mar 2014 20:52:40 +0200
Subject: [PATCH] SystemUI: Fix missing top task in recent panel

If preloadRecentTasksList was not called or was called too
late and RecentsActivty already started and last top activity has
excludeFromRecents flag setted it might be missing
from recent application list.

atken from cm11

Change-Id: Ia04913e7a89d51689f933cbc9ff3a84092fdb1cf
---
 core/java/android/app/ActivityManager.java                 |  5 +++++
 .../src/com/android/systemui/recent/RecentTasksLoader.java | 14 +++++++++++++-
 .../java/com/android/server/am/ActivityManagerService.java |  6 +++++-
 3 files changed, 23 insertions(+), 2 deletions(-)

diff --git a/core/java/android/app/ActivityManager.java b/core/java/android/app/ActivityManager.java
index 7ca3459..9866ac3 100644
--- a/core/java/android/app/ActivityManager.java
+++ b/core/java/android/app/ActivityManager.java
@@ -568,6 +568,11 @@ public class ActivityManager {
     public static final int RECENT_IGNORE_UNAVAILABLE = 0x0002;
 
     /**
+     * @hide
+     */
+    public static final int RECENT_DO_NOT_COUNT_EXCLUDED = 0x0004;
+
+    /**
      * Return a list of the tasks that the user has recently launched, with
      * the most recent being first and older ones after in order.
      *
diff --git a/packages/SystemUI/src/com/android/systemui/recent/RecentTasksLoader.java b/packages/SystemUI/src/com/android/systemui/recent/RecentTasksLoader.java
index c714d8b..30c4754 100644
--- a/packages/SystemUI/src/com/android/systemui/recent/RecentTasksLoader.java
+++ b/packages/SystemUI/src/com/android/systemui/recent/RecentTasksLoader.java
@@ -439,12 +439,16 @@ public class RecentTasksLoader implements View.OnTouchListener {
                 mContext.getSystemService(Context.ACTIVITY_SERVICE);
 
                 final List<ActivityManager.RecentTaskInfo> recentTasks =
-                        am.getRecentTasks(MAX_TASKS, ActivityManager.RECENT_IGNORE_UNAVAILABLE);
+                        am.getRecentTasks(MAX_TASKS, ActivityManager.RECENT_IGNORE_UNAVAILABLE
+                                | ActivityManager.RECENT_WITH_EXCLUDED
+                                | ActivityManager.RECENT_DO_NOT_COUNT_EXCLUDED);
                 int numTasks = recentTasks.size();
                 ActivityInfo homeInfo = new Intent(Intent.ACTION_MAIN)
                         .addCategory(Intent.CATEGORY_HOME).resolveActivityInfo(pm, 0);
 
                 boolean firstScreenful = true;
+                boolean loadOneExcluded = true;
+
                 ArrayList<TaskDescription> tasks = new ArrayList<TaskDescription>();
 
                 // skip the first task - assume it's either the home screen or the current activity.
@@ -462,6 +466,7 @@ public class RecentTasksLoader implements View.OnTouchListener {
 
                     // Don't load the current home activity.
                     if (isCurrentHomeActivity(intent.getComponent(), homeInfo)) {
+                        loadOneExcluded = false;
                         continue;
                     }
 
@@ -470,6 +475,13 @@ public class RecentTasksLoader implements View.OnTouchListener {
                         continue;
                     }
 
+                    if (!loadOneExcluded && (recentInfo.baseIntent.getFlags()
+                            & Intent.FLAG_ACTIVITY_EXCLUDE_FROM_RECENTS) != 0) {
+                        continue;
+                    }
+
+                    loadOneExcluded = false;
+
                     TaskDescription item = createTaskDescription(recentInfo.id,
                             recentInfo.persistentId, recentInfo.baseIntent,
                             recentInfo.origActivity, recentInfo.description);
diff --git a/services/java/com/android/server/am/ActivityManagerService.java b/services/java/com/android/server/am/ActivityManagerService.java
index 9201b1d..688173b 100644
--- a/services/java/com/android/server/am/ActivityManagerService.java
+++ b/services/java/com/android/server/am/ActivityManagerService.java
@@ -6858,7 +6858,11 @@ public final class ActivityManagerService extends ActivityManagerNative
                     }
                     
                     res.add(rti);
-                    maxNum--;
+                    if ((tr.intent.getFlags() & Intent.FLAG_ACTIVITY_EXCLUDE_FROM_RECENTS) == 0
+                            || (flags & ActivityManager.RECENT_DO_NOT_COUNT_EXCLUDED) == 0
+                            || i == 0) {
+                        maxNum--;
+                    }
                 }
             }
             return res;
-- 
1.9.1

